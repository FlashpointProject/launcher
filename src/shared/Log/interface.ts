export enum LogLevel {
  TRACE,
  DEBUG,
  INFO,
  WARN,
  ERROR,
  SILENT,
}

/** A log entry _before_ it is added to the main log */
export type ILogPreEntry = {
  /** Name of the source of the log entry (name of what added the log entry) */
  source: string;
  /** Content of the log entry */
  content: string;
}

/** A log entry from the main log */
export type ILogEntry = ILogPreEntry & {
  /** Timestamp of when the entry was added to the main's log */
  timestamp: number;
  /** Level of the log, 0-5, Trace, Info, Warn, Error, Fatal */
  logLevel: LogLevel;
}
