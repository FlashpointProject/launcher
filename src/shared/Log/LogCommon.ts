import { padStart } from '../Util';
import { ILogEntry } from './interface';

export const timeChars = 11; // "[HH:MM:SS] "
const sourceChars = 19; // "Background Services" (sometimes used with +2 to add the length of ": ")

/** Create a HTML string of a number of entries */
export function stringifyLogEntries(entries: ILogEntry[], filter: { [key: string]: boolean } = {}): string {
  let str = '';
  let prevEntry: ILogEntry = { source: '', content: '', timestamp: 0 };
  for (let i = 0; i < entries.length; i++) {
    const entry = entries[i];
    const entryFilter = filter[entry.source];
    if (entryFilter === true || entryFilter === undefined) {
      str += `<span class="log__time-stamp">[${formatTime(new Date(entry.timestamp))}]</span> `;
      if (entry.source) {
        str += (entry.source !== prevEntry.source) ?
              `<span class="log__source log__source--${getClassModifier(entry.source)}">${padStart(escapeHTML(entry.source), sourceChars)}:</span> ` :
              ' '.repeat(sourceChars + 2);
      }
      str += padLines(escapeHTML(entry.content), timeChars + sourceChars + 2);
      str += '\n';
    }
    prevEntry = entry;
  }
  return str;
}

/** Create a HTML string of a number of entries */
export function stringifyLogEntriesRaw(entries: ILogEntry[], filter: { [key: string]: boolean } = {}): string {
  let str = '';
  let prevEntry: ILogEntry = { source: '', content: '', timestamp: 0 };
  for (let i = 0; i < entries.length; i++) {
    const entry = entries[i];
    const entryFilter = filter[entry.source];
    if (entryFilter === true || entryFilter === undefined) {
      str += `[${(new Date(entry.timestamp)).toLocaleString()}] `;
      if (entry.source) {
        str += (entry.source !== prevEntry.source) ?
              `${padStart(escapeHTML(entry.source), sourceChars)} | ` :
              ' '.repeat(sourceChars + 2);
      }
      str += padLines(entry.content, timeChars + sourceChars + 2);
      str += '\n';
    }
    prevEntry = entry;
  }
  return str;
}

/** Formats a date to a string in the format "HH:MM:SS" */
export function formatTime(date: Date): string {
  return (
    ('0'+date.getHours()  ).slice(-2)+':'+
    ('0'+date.getMinutes()).slice(-2)+':'+
    ('0'+date.getSeconds()).slice(-2)
  );
}

/** Mak a string safe to use as HTML content (only safe if used as "text" between tags, not inside a tag) */
export function escapeHTML(str: string): string {
  return str.replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;');
}

/**
 * Create a CSS class "modifier" name from the name of a log entry source
 * (it just makes it lower-case, only alphabetical characters and replaces all spaces with "-")
 */
function getClassModifier(source: string): string {
  return (
    source
      .toLowerCase()
      .replace(/ /g, '-')
      .replace(/[^a-z-]/gi, '') // (Only allow a-z and "-")
  );
}

/** Pad all lines (except for the first one) by a number of spaces */
export function padLines(text: string, padding: number): string {
  return text.replace(/\n/g, '\n'+' '.repeat(padding));
}
