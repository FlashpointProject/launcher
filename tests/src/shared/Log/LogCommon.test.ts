import { ILogEntry } from '@shared/Log/interface';
import { stringifyLogEntries, stringifyLogEntriesRaw } from '@shared/Log/LogCommon';

describe('Log Functions', () => {
  test('Stringify Log Entries', () => {
    const entryDate = new Date();
    entryDate.setHours(12);
    entryDate.setMinutes(34);
    entryDate.setSeconds(56);
    const entry: ILogEntry = {
      source: 'Game Launcher',
      content: 'TestSuccess',
      timestamp: entryDate.valueOf()
    };

    const stringData = JSON.stringify(stringifyLogEntries([entry]));
    expect(stringData).toMatchSnapshot();
  });

  test('Stringify Log Entries Raw', () => {
    const entryDate = new Date('01-01-2000');
    entryDate.setHours(12);
    entryDate.setMinutes(34);
    entryDate.setSeconds(56);
    const entry: ILogEntry = {
      source: 'Game Launcher',
      content: 'TestSuccess',
      timestamp: entryDate.valueOf()
    };

    const stringData = JSON.stringify(stringifyLogEntriesRaw([entry]));
    expect(stringData).toMatchSnapshot();
  });
});