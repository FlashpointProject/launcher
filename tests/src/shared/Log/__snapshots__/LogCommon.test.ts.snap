// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log Functions Stringify Log Entries 1`] = `"\\"<span class=\\\\\\"log__time-stamp\\\\\\">[12:34:56]</span> <span class=\\\\\\"log__source log__source--game-launcher\\\\\\">      Game Launcher:</span> TestSuccess\\\\n\\""`;

exports[`Log Functions Stringify Log Entries Raw 1`] = `"\\"[1/1/2000, 12:34:56 PM]       Game Launcher | TestSuccess\\\\n\\""`;
